Mathlib/Algebra/BigOperators/Fin.lean

```diff
@@ -0,0 +1,383 @@
+  rw [prod_univ_add, Fintype.prod_eq_one _ hf, mul_one]
+  rfl
+#align fin.prod_trunc Fin.prod_trunc
+#align fin.sum_trunc Fin.sum_trunc
+
+section PartialProd
+
+variable [Monoid α] {n : ℕ}
+
+/-- For `f = (a₁, ..., aₙ)` in `αⁿ`, `partial_prod f` is `(1, a₁, a₁a₂, ..., a₁...aₙ)` in `αⁿ⁺¹`. -/
+@[to_additive "For `f = (a₁, ..., aₙ)` in `αⁿ`, `partial_sum f` is\n
+`(0, a₁, a₁ + a₂, ..., a₁ + ... + aₙ)` in `αⁿ⁺¹`."]
+def partialProd (f : Fin n → α) (i : Fin (n + 1)) : α :=
+  ((List.ofFn f).take i).prod
+#align fin.partial_prod Fin.partialProd
+#align fin.partial_sum Fin.partialSum
+
+@[to_additive (attr := simp)]
+theorem partialProd_zero (f : Fin n → α) : partialProd f 0 = 1 := by simp [partialProd]
+#align fin.partial_prod_zero Fin.partialProd_zero
+#align fin.partial_sum_zero Fin.partialSum_zero
+
+@[to_additive]
+theorem partialProd_succ (f : Fin n → α) (j : Fin n) :
+    partialProd f j.succ = partialProd f (Fin.castSucc j) * f j := by
+  simp [partialProd, List.take_succ, List.ofFnNthVal, dif_pos j.is_lt, ← Option.coe_def]
+#align fin.partial_prod_succ Fin.partialProd_succ
+#align fin.partial_sum_succ Fin.partialSum_succ
+
+@[to_additive]
+theorem partialProd_succ' (f : Fin (n + 1) → α) (j : Fin (n + 1)) :
+    partialProd f j.succ = f 0 * partialProd (Fin.tail f) j := by
+  simp [partialProd]
+  rfl
+#align fin.partial_prod_succ' Fin.partialProd_succ'
+#align fin.partial_sum_succ' Fin.partialSum_succ'
+
+@[to_additive]
+theorem partialProd_left_inv {G : Type _} [Group G] (f : Fin (n + 1) → G) :
+    (f 0 • partialProd fun i : Fin n => (f i)⁻¹ * f i.succ) = f :=
+  funext fun x => Fin.inductionOn x (by simp) fun x hx => by
+    simp only [coe_eq_castSucc, Pi.smul_apply, smul_eq_mul] at hx⊢
+    rw [partialProd_succ, ← mul_assoc, hx, mul_inv_cancel_left]
+#align fin.partial_prod_left_inv Fin.partialProd_left_inv
+#align fin.partial_sum_left_neg Fin.partialSum_left_neg
+
+-- Porting note:
+-- 1) Changed `i` in statement to `(Fin.castLt i (Nat.lt_succ_of_lt i.2))` because of
+--    coersion issues. Might need to be fixed later.
+-- 2) The current proof is really bad! It should be redone once `assoc_rw` is
+--    implemented and `rw` knows that `i.succ = i + 1`.
+-- 3) The original Mathport output was:
+--   cases' i with i hn
+--   induction' i with i hi generalizing hn
+--   · simp [← Fin.succ_mk, partialProd_succ]
+--   · specialize hi (lt_trans (Nat.lt_succ_self i) hn)
+--     simp only [mul_inv_rev, Fin.coe_eq_castSucc, Fin.succ_mk, Fin.castSucc_mk, smul_eq_mul,
+--       Pi.smul_apply] at hi ⊢
+--     rw [← Fin.succ_mk _ _ (lt_trans (Nat.lt_succ_self _) hn), ← Fin.succ_mk]
+--     simp only [partialProd_succ, mul_inv_rev, Fin.castSucc_mk]
+--     assoc_rw [hi, inv_mul_cancel_left]
+@[to_additive]
+theorem partialProd_right_inv {G : Type _} [Group G] (g : G) (f : Fin n → G) (i : Fin n) :
+    ((g • partialProd f) (Fin.castLt i (Nat.lt_succ_of_lt i.2)))⁻¹ *
+    (g • partialProd f) i.succ = f i := by
+  rcases i with ⟨i, hn⟩
```

Comment:
This theorem was a nightmare to prove. Lots of things that should be def-eq aren't recoginsed as such by `rw` (e.g. `n + 1` vs `n.succ`). Moreover `assoc_rw` would be really helpful.

Could somebody look at the proof and decide if there is anything that could be done to avoid non-terminal `simp` and all the `change`?

Reply:
Also the statement has changed. In Lean3 it was `↑i` instead of `(Fin.castLt i (Nat.lt_succ_of_lt i.2))`. I changed that because Lean4 got the conversion wrong, could that be fixed too?

Reply:
I see, Lean 4 coerces via the path `Fin n → ℕ → Fin (n + 1)` which results in different defeqs than `castLt`. Although I can't actually find where the coercion comes from in mathlib 3...

Reply:
Nope, turns out that Lean 3 finds the same coercion (`@coe_trans (fin n) ℕ (fin (n + 1)) nat.cast_coe (fin.fin_to_nat n)`). So I think it's best to keep the statement with a coercion.

Reply:
Thanks for writing the extensive porting note.
Let's merge this PR now. We can revisit this issue later.